# Calculating Accrued Interest

## Preparing the notebook

### Importing libraries, modules, And functions

 As in Chapter One of the volume, modules that are included in the standard Python library are imported. When necessary, other modules or libraries are installed before they are imported.$^{1}$  

```
import os
import sys
import requests
from datetime import , date
from types import ModuleType
import calendar

try:
    from IPython.display import Markdown, display
except:
    !pip -q install IPython
    from IPython.display import Markdown, display

try:
    import numpy as np
except:
    !pip -q install numpy
    import numpy as np

try:
    import pandas as pd
except:
    !pip -q install pandas
    import pandas as pd

try:
    from dateutil.relativedelta import relativedelta
except:
    !pip -q install python-dateutil
    from dateutil.relativedelta import relativedelta

```

---
$^{1}$<a href='https://patrickjhess.github.io/Introduction-To-Python-For-Financial-Python/Control_Statements.html#the-try-and-except'>try and except statements</a>.

In [1]:
# Import OS to interact with local computer operating system
import os
import sys
import requests
from types import ModuleType
# Import the datetime and date class from the datetime module for working with dates.
from datetime import date
# Last calendar day of the month and day of the week for first day
import calendar

try:
    from IPython.display import Markdown, display
except:
    !pip -q install IPython
    from IPython.display import Markdown, display

# Import the numpu library for data manipulation and analysis, aliased as np
try:
    import numpy as np
except:
    !pip -q install numpy
    import numpy as np
# Import the pandas library for data manipulation and analysis, aliased as pd.
try:
    import pandas as pd
except:
    !pip -q install pandas
    import pandas as pd

# Import the relativedelta class from dateutil for advanced date calculations.
try:
    from dateutil.relativedelta import relativedelta
except:
    !pip -q install python-dateutil
    from dateutil.relativedelta import relativedelta

### Adding a custom module and importing functions



 Like Chapter One the custom module <font color='green'>module_basic_concepts_fixed_income</font> custom module  is accessed from Dropbox and named <font color='green'>basic_concepts_fixed_income</font>.  As a reminder, the module is accessible in the notebook's memory, but is not added to a drive.  One function, <font color='green'>create_workbood</font>, is added to the notebook. Any functions that are includced in the module are avaialable to <font color='green'>create_workbook</font>.


```
from basic_income_module(create_workbook)
```


*    <font color='green'>create_workbook()</font> [view here](https://patrickjhess.github.io/Imported-Functions/create_workbook.html#create-workbook-is-a-helper-function-that-creates-excel-workbooks-from-dataframes-created-by-financial-python)





In [2]:
# Define the URL of the Python module to be downloaded from Dropbox.
# The 'dl=1' parameter in the URL forces a direct download of the file content.
url= 'https://www.dropbox.com/scl/fi/4y5hjxlfphh1ngvbgo77q/\
module_-basic_concepts_fixed_income.py?rlkey=6oxi7mgka42veaat79hcv8boz&st=87sztshr&dl=1'
module_name='basic_concepts_fixed_income'
# Send an HTTP GET request to the URL and store the server's response.
try:
  response=requests.get(url)
  # Raise an exception for bad status codes (like 404 Not Found)
  response.raise_for_status()
  module= ModuleType(module_name)
  #Code contained in response.text executed
  exec(response.text, module.__dict__)
  # Module added to sys
  sys.modules[module_name]=module
except requests.exceptions.RequestException as e:
    print(f"❌ Error: Could not fetch module from URL. {e}")
except Exception as e:
    print(f"❌ Error: Failed to execute or import the module. {e}")

# Now that 'basic_concepts_fixed_income' exists in the notebook, import the specific functions
from basic_concepts_fixed_income import(create_workbook)

## Creating the DataFrame bond_data
US Treasury notes and bonds data, for settlement on January 21, 2025,  obtained from Fidelity, an Excel workbook, is downloaded from DropBox using the Panda's <font color='green'>read_excel()</font> method.$^{2}$

 Three arguments are passed to the method.



1.    The URL address (<font color='green'>url</font>) is required.
2.    Assigning the index column is optional. The maturity dates, located in the first column, are set as the index of the DataFrame  (<font color='green'>index_col='Maturity Date'</font>).
3.    The worksheet name defaults to the first worksheet. The worksheet's name, 'Fidelity Data', is the assigned sheet (<font color='green'>sheet_name='Fidelity Data'</font>)

The <font color='green'>display()</font>  function shows the first and last five rows confirming that data has been successfully accessed.

---

 $^{2}$ <a href='https://patrickjhess.github.io/Introduction-To-Python-For-Financial-Python/An_Introduction_To_Pandas.html#dataframes-csv-and-excel-files'>Pandas method read_excel</a>

In [3]:
#The full file path.
url='https://www.dropbox.com/scl/fi/lgnaj41bt8o9sv5a63rr1/\
bond_data_jan21_2025.xlsx?rlkey=twjzkcqo0g2ahvot78518ti4x&st=ihc5feog&dl=1'
print(f"Attempting to load data from:\n {url}")

#Load the data from Excel, using the first column as the index.
try:
    bond_data = pd.read_excel(url, index_col='Maturity Date',sheet_name='Fidelity Data')

    # Display the first and last 5 rows of the loaded DataFrame to verify it worked..
    display(bond_data)

except FileNotFoundError:
    print("\nERROR: File not found.")
    print("Please check that the 'folder' and 'file' variables are spelled correctly'\
' and that the file exists in that location.")

Attempting to load data from:
 https://www.dropbox.com/scl/fi/lgnaj41bt8o9sv5a63rr1/bond_data_jan21_2025.xlsx?rlkey=twjzkcqo0g2ahvot78518ti4x&st=ihc5feog&dl=1


Description  Coupon  \
Maturity Date                                                              
2025-01-28     UNITED STATES TREAS BILLS ZERO CPN       0.000...   0.000   
2025-01-30     UNITED STATES TREAS BILLS ZERO CPN       0.000...   0.000   
2025-01-31     UNITED STATES TREAS SER U-2025          1.3750...   1.375   
2025-01-31     UNITED STATES TREAS SER AW-2025         4.1250...   4.125   
2025-01-31     UNITED STATES TREAS SER G-2025          2.5000...   2.500   
...                                                          ...     ...   
2029-11-30     UNITED STATES TREAS SER AG-2029         4.1250...   4.125   
2029-11-30     UNITED STATES TREAS SER S-2029          3.8750...   3.875   
2029-12-31     UNITED STATES TREAS SER T-2029          3.8750...   3.875   
2029-12-31     UNITED STATES TREAS SER AH-2029         4.3750...   4.375   
2030-01-31     UNITED STATES TREAS SER G-2030          3.5000...   3.500   

               Price Bid  Price Ask  Bid Size  Ask Size  
Maturity Date                                            
2025-01-28        99.929     99.930    100000    100000  
2025-01-30        99.906     99.907     40000     40000  
2025-01-31        99.921     99.934     60000     60000  
2025-01-31        99.988     99.997     60000     60000  
2025-01-31        99.953     99.965     60000     60000  
...                  ...        ...       ...       ...  
2029-11-30        98.910     98.914    100000    100000  
2029-11-30        97.763     97.782     40000     40000  
2029-12-31        97.734     97.738     65000     65000  
2029-12-31        99.988     99.989      7000      7000  
2030-01-31        95.994     96.011     16000     16000  

[304 rows x 6 columns]

## Validating dates and converting <font color='green'>datetime</font> to <font color='green'>date</font>.
In the *Manipulating Dates* section, we noted that the datetime library creates two distinct, but related, object types. For calculations like accrued interest or bond payment dates, <font color='green'>date</font> objects are preferable to  <font color='green'>datetime</font> objects because they only contain the necessary year, month, and day attributes.

<font color='green'>datetime</font> objects include unnecessary time attributes (hour, second, microsecond) that can complicate comparisons between two dates with the same year, month, and day. To prevent such issues, all <font color='green'>datetime</font> objects are converted to <font color='green'>date</font> objects.

The <font color='green'>validate_date</font> function ensures its argument is either a  <font color='green'>datetime</font> or  <font color='green'>date</font>, raising an error if it is neither. It also handles the conversion of any  <font color='green'>datetime</font> instance to a  <font color='green'>date</font> object.

In [4]:
def validate_date(datetime_obj):
  """
  Validates input is a date/datetime and returns a date object.

  This helper function checks if the input is either a
  datetime.datetime or datetime.date object. If it's a
  datetime.datetime, it converts it to a datetime.date.
  It always returns a datetime.date object.

  Args:
    datetime_obj (datetime.date or datetime.datetime):
      The input to validate and convert.

  Returns:
    datetime.date: The resulting date object.

  Raises:
    TypeError: If the input is not a datetime.date or
               datetime.datetime object.
  """
  from datetime import datetime, date

  #Validation ---

  # Check if the object is of the expected type (date or datetime)
  if not isinstance(datetime_obj, (datetime, date)):
      raise TypeError("Input must be a datetime or date object.")

  #Standardization ---

  # If the object is a datetime, convert it to a date.
  # This makes the return type consistent.
  if isinstance(datetime_obj, datetime):
    datetime_obj = datetime_obj.date()

  # Return the object, which is now guaranteed to be a date object.
  return datetime_obj

## Illustrating accrued interest with five unique maturity dates


The <font color='green'>loc</font> attribute is used to select bonds based on the five maturity dates. Because the <font color='green'>sample_bonds</font> DataFrame contains multiple bonds with the same maturity date, the Pandas <font color='green'>duplicated</font> method is employed to obtain five unique bonds, each corresponding to a distinct maturity date. When <font color='green'>keep</font> is set to 'first', `True` is assigned to the first occurrence of a duplicated value and `False` to subsequent duplicates. Consequently, <font color='green'>loc</font> only returns the rows where <font color='green'>duplicated()</font> is `True`, yielding the five bonds with unique maturity dates.$^{3}$
```
maturity_dates = [
    date(2025, 2, 28),
    date(2025, 7, 15),
    date(2025, 8, 31),
    date(2026, 1, 15),
    date(2028, 2, 29),
]
sample_bonds=bond_data.loc[maturity_dates]
unique_bond= ~all_sample_bonds.index.duplicated(keep='first')
display(unique_bonds)
five_bonds=sample_bonds.loc[unique_bonds]
display(five_bonds)


```

$^{3}$ <a href='https://patrickjhess.github.io/Introduction-To-Python-For-Financial-Python/An_Introduction_To_Pandas.html#pandas-duplicated-method'>The Pandas Method duplicated</a>

In [8]:
# --- Define a specific list of upcoming bond maturity dates to analyze ---
# These dates were selected for a targeted review of bonds maturing in
# late 2025 and early 2026.
maturity_dates = [
    date(2025, 2, 28),
    date(2025, 7, 15),
    date(2025, 8, 31),
    date(2026, 1, 15),
    date(2028, 2, 29),
]

# Use the .loc indexer to perform a label-based lookup.
# This retrieves all rows where the DataFrame's index exactly matches the dates in the list.
sample_bonds = bond_data.loc[maturity_dates]
# ~ negates the value of duplicated..the maturity is unique
# keep the first instance that is not a duplicate...a unique index is unmodified
unique_bonds= ~sample_bonds.index.duplicated(keep='first')
# display the value of the series unique_bonds
display(unique_bonds)
five_bonds=sample_bonds.loc[unique_bonds]

# Render the resulting 'five_bonds' DataFrame as a formatted table in the output.
display(five_bonds)

array([ True, False, False,  True,  True, False, False,  True,  True,
       False])

Description  Coupon  \
Maturity Date                                                              
2025-02-28          UNITED STATES TREAS NOTE 1.12500% 02/28/2025   1.125   
2025-07-15     UNITED STATES TREAS SER AQ-2025         3.0000...   3.000   
2025-08-31     UNITED STATES TREAS SER AC-2025         0.2500...   0.250   
2026-01-15     UNITED STATES TREAS SER AJ-2026         3.8750...   3.875   
2028-02-29     UNITED STATES TREAS SER V-2028          4.0000...   4.000   

               Price Bid  Price Ask  Bid Size  Ask Size  
Maturity Date                                            
2025-02-28        99.671     99.684     60000     60000  
2025-07-15        99.394     99.406     60000     60000  
2025-08-31        97.605     97.610     60000     60000  
2026-01-15        99.671     99.688     40000     40000  
2028-02-29        99.099     99.118     40000     40000

## <font color='green'>Application:Access Bonds With Specific Characteristics</font>

<div style="
    border-left: 12px solid green;
    line-height: 1.5;
    padding: 15px">
<br>

Access the bonds that mature at the end of April and October 2025. Create a DataFrame of the bonds with a coupon greater than 3.5. For hints, see [Chapter Two Hints:Access Bonds With Specific Characteristics](https://patrickjhess.github.io/Hints-Results/Chapter_Two_Hints.html#access-bonds-with-specific-characteristics), and check the [expected results here](https://patrickjhess.github.io/Hints-Results/Chapter_Two_Results.html#access-bonds-with-specific-characteristics).

<br>
</div>


## Scheduled payment dates

To calculate a bond's accrued interest, the scheduled payment dates must be known. Because the bond's maturity date is a scheduled payment date, all other scheduled payment dates can be determined by working backward from the maturity date, using the time interval between payments. The <font color='green'>scheduled_pay_dates</font> function performs this calculation.

The function uses a <font color='green'>while</font> loop, starting at the maturity date and decrementing backward toward the settlement date. In each iteration, a scheduled <font color='green'>pay_date</font> is appended to the dates <font color='green'>list</font>.

The time between scheduled payments is represented by <font color='green'>num_months</font> (e.g., six for a semi-annual bond). The <font color='green'>pay_date</font> value  is reduced by this interval in each iteration using  <font color='green'>relativedelta</font>. The variable <font color='green'>isMontEnd</font> is set using <font color='green'>calendar.monthrange</font> (as demonstrated in the *Manipulating Dates* notebook) to determine if the maturity is a month-end date. If the maturity is a month-end, the corresponding iteration's <font color='green'>pay_date</font> is assigned the last day of its month.

The list of dates is initially generated in reverse order because the calculation starts from the maturity date. To correct this sequence, the list is then reversed by using the <font color='green'>[::-1]</font> slice notation (starting at the end and moving backwards).$^{4}$

The following example demonstrates the function for a settlement date of January 21, 2025, and a maturity date of February 28, 2030:

```py
settlement = date(2025, 1, 21)
maturity = date(2030, 2, 28)
dates = []
pay_date = maturity
is_month_end = maturity.day == calendar.monthrange(maturity.year, maturity.month)[1]
num_months=6
dates = []
pay_date = maturity

#Loop backward from the maturity date

while pay_date > settlement:
  dates.append(pay_date)
      
  # Decrement by the frequency
  pay_date -= relativedelta(months=num_months)

  # Handle Month End Logic
  if is_month_end:
    last_day = calendar.monthrange(pay_date.year, pay_date.month)[1]
    pay_date = date(pay_date.year, pay_date.month, last_day)
dates[::-1]
```
$^{4}$ see [see A Quick Introduction To Lists](https://)



In [9]:
settlement = date(2025, 1, 21)
maturity = date(2030, 2, 28)
dates = []
pay_date = maturity
is_month_end = maturity.day == calendar.monthrange(maturity.year, maturity.month)[1]
num_months=6
dates = []
pay_date = maturity

#Loop backward from the maturity date

while pay_date > settlement:
  dates.append(pay_date)

  # Decrement by the frequency
  pay_date -= relativedelta(months=num_months)

  # Handle Month End Logic
  if is_month_end:
    last_day = calendar.monthrange(pay_date.year, pay_date.month)[1]
    pay_date = date(pay_date.year, pay_date.month, last_day)
dates[::-1]

[datetime.date(2025, 2, 28),
 datetime.date(2025, 8, 31),
 datetime.date(2026, 2, 28),
 datetime.date(2026, 8, 31),
 datetime.date(2027, 2, 28),
 datetime.date(2027, 8, 31),
 datetime.date(2028, 2, 29),
 datetime.date(2028, 8, 31),
 datetime.date(2029, 2, 28),
 datetime.date(2029, 8, 31),
 datetime.date(2030, 2, 28)]

## The <font color='green'>schedule_pay_dates</font> function

The <font color='green'>scheduled_pay_dates</font> function requires the bond's maturity date (a <font color='green'>datetime</font> or <font color='green'>date</font> object) as its argument. By default, the settlement date is the current date, and the payment frequency is set to '2' for semi-annual payments. The <font color='green'>validated_date</font> function is used to validate the provided maturity and settlement dates

In [10]:
def scheduled_pay_dates(maturity,settlement=None,freq=2):
  '''
    Generates a chronological list of coupon payment dates from settlement to maturity.
    The function calculates dates backward from the maturity date based on the
    specified frequency. It handles standard bond market "end-of-month" logic:
    if the maturity date is the last day of a month, all preceding coupon payments
    are snapped to the last day of their respective months.
    Args:
        maturity (datetime.date): The final maturity date of the bond.
            Accepts a date object.
        settlement (datetime.date, optional): The settlement date (start of analysis).
            Coupons falling before this date are excluded. Defaults to date.today().
        freq (int, optional): The number of coupon payments per year.
            Accepted values:
            * 1: Annual, 2: Semi-Annual (Default), 4: Quarterly, 12: Monthly

    Returns:
        list[datetime.date]: A list of coupon dates sorted chronologically
        (earliest to latest), ending with the maturity date..
  '''
  from datetime import datetime,date
  import calendar
  from dateutil.relativedelta import relativedelta

  #Validate the data- maturity, coupon, settlement, freq
  #maturity
  maturity=validate_date(maturity)

  #settlement
  if settlement is None:
      settlement = date.today()
  else:
      settlement=validate_date(settlement)
  #freq
  if int(freq) not in [1,2,4,12]:
      display(md(f"### ⚠️  your assigned freq {freq} it must be (1, 2, 4, or 12)\
     \n     semi-annual assumed (2)."))
      freq=int(2)

  # Calculate the number of months between each coupon payment.
  num_months=int(12/freq)

  #Need to check for month_end
  is_month_end = maturity.day == calendar.monthrange(maturity.year, maturity.month)[1]

  dates = []
  pay_date = maturity

  #Loop backward from the maturity date

  while pay_date > settlement:
    dates.append(pay_date)

    # Decrement by the frequency
    pay_date -= relativedelta(months=num_months)

    # Handle Month End Logic
    if is_month_end:
      last_day = calendar.monthrange(pay_date.year, pay_date.month)[1]
      pay_date = date(pay_date.year, pay_date.month, last_day)

  # Return chronologically (sliced backward)
  return dates[::-1]

## The next and previous scheduled payment dates

The <font color='green'>scheduled_pay_dates</font> function is used to determine the next and previous schedule payments.

Specifically:

1. The next scheduled payment date is the first date greater than or equal to the settlement date. It should be the first value in the list returned by <font color='green'>scheduled_pay_dates</font>.  
2. The last scheduled payment date is calculated by taking the next scheduled payment date and subtracting the number of months between payments, while also accounting for the month-end adjustment.

This calculation is demonstrated using the  <font color='green'>five_bonds</font> DataFrame.

In [11]:
settlement=date(2025,1,21)

# Bond maturities on the index of the DataFrame
# Convert Pandas Timestamps to date object
maturities=five_bonds.index.date

#Iterate through maturities
for maturity in maturities:
  # First element of the list
  next_coupon_date=scheduled_pay_dates(maturity,settlement=settlement,freq=2)[0]
  # relativedelta subtracts six months
  previous_coupon_date=next_coupon_date-relativedelta(months=6)
  #month-end adjustment if needed
  is_month_end=maturity.day==calendar.monthrange(maturity.year, maturity.month)[1]
  if is_month_end:
    last_day=calendar.monthrange(previous_coupon_date.year, previous_coupon_date.month)[1]
    previous_coupon_date=date(previous_coupon_date.year, previous_coupon_date.month,last_day)
  display(f'maturity {maturity}--previous {previous_coupon_date}--next {next_coupon_date}')

'maturity 2025-02-28--previous 2024-08-31--next 2025-02-28'

'maturity 2025-07-15--previous 2025-01-15--next 2025-07-15'

'maturity 2025-08-31--previous 2024-08-31--next 2025-02-28'

'maturity 2026-01-15--previous 2025-01-15--next 2025-07-15'

'maturity 2028-02-29--previous 2024-08-31--next 2025-02-28'

## <font color='green'>Application:Calculate Days For Accrued Interest</font>

<div style="
    border-left: 12px solid green;
    line-height: 1.5;
    padding: 15px">
<br>

Assume a bond matures on January 15$^{th}$ 2030 and the settlement date is May 1$^{st}$ 2024.  Calculate the number of days since the last coupon payment and the number of days between the last and next coupon payment. see [Chapter Two Calculate Days For Accrued Interest](https://patrickjhess.github.io/Hints-Results/Chapter_Two_Hints.html#calculate-days-between-coupon-payment-and-settlement-dates), and check the [expected results here](https://patrickjhess.github.io/Hints-Results/Chapter_Two_Results.html#calculate-days-between-coupon-payment-and-settlement-dates).

<br>
</div>

## Calculating accrued interest



Calculating the number of days between or since payments is typically a straightforward subtraction of dates. The "Actual/Actual," "Actual/365," and "Actual/360" day-count conventions are simple to compute-a difference between two date objects; the "30/360" convention requires a more specialized approach dealt with  by the function <font color='green'>convert_30_360</font>.

```py
def convert_30_360(start_date, end_date):
    from datetime import date
    import calendar

    start_date = validate_date(start_date)
    end_date = validate_date(end_date)

    d1, m1, y1 = start_date.day, start_date.month, start_date.year
    d2, m2, y2 = end_date.day, end_date.month, end_date.year

    def is_last_day_of_feb(d: date) -> bool:
        return d.month == 2 and d.day == calendar.monthrange(d.year, d.month)[1]

    if is_last_day_of_feb(start_date) and is_last_day_of_feb(end_date):
        d2 = 30
    if d1 == 31 or is_last_day_of_feb(start_date):
        d1 = 30

    if d2 == 31 and d1 == 30:
        d2 = 30

    day_count = (y2 - y1) * 360 + (m2 - m1) * 30 + (d2 - d1)
    return day_count
```

Here are examples illustrating the differences in day count using the 30/360 convention:

* **January 21, 2025, to August 31, 2024:**  
  * 30/360 convention: 140 days  
  * Actual: 143 days  
* **February 28, 2025, to August 31, 2024:**  
  * 30/360 convention: 180 days  
  * Actual: 181 days



In [ ]:
def convert_30_360(start_date, end_date):
  """
  Calculates the number of days between two dates using the 30U/360
  (US Bond Basis) convention.

  Args:
      start_date: The beginning date of the period.
      end_date: The ending date of the period.

  Returns:
      The number of days calculated using the 30U/360 convention.
  """
  from datetime import date
  import calendar

  #Validate data
  #start_date
  start_date=validate_date(start_date)

  #end_date
  end_date=validate_date(end_date)

  d1, m1, y1 = start_date.day, start_date.month, start_date.year
  d2, m2, y2 = end_date.day, end_date.month, end_date.year

  # Helper to check if a date is the last day of February
  def is_last_day_of_feb(d: date) -> bool:
      return d.month == 2 and d.day == calendar.monthrange(d.year, d.month)[1]

  # --- Apply the 30U/360 rules ---

  #  If both dates are the last day of Feb, change end_date's day to 30.
  if is_last_day_of_feb(start_date) and is_last_day_of_feb(end_date):
      d2 = 30

  #  If start_date is 31 or last day of Feb, change its day to 30.
  if d1 == 31 or is_last_day_of_feb(start_date):
      d1 = 30

  # If end_date is 31 AND start_date's day was changed to 30, change end_date's day to 30.
  if d2 == 31 and d1 == 30:
      d2 = 30

  # --- Perform the final calculation ---
  day_count = (y2 - y1) * 360 + (m2 - m1) * 30 + (d2 - d1)

  return day_count

## The accrued_interest function

Accrued interest is determined by the annual coupon, adjusted by the payment frequency, and then multiplied by a fraction representing the time elapsed since the last payment. This time fraction is calculated as the ratio of days since the last payment to the total days between payments, with the specific calculation method depending on the day-count convention.


In [14]:
def accrued_interest(maturity, coupon, day_type='Actual/Actual', settlement=None, freq=2):
    """
      Returns the accrued interest for a bond.  Returns the accrued interest for a bond.

    Args:
        maturity (datetime): The maturity date of the bond.
        coupon (float): The annual coupon rate (e.g., 0.05 for 5%).
        day_types:
            Actual/Actual, Actual/365, Actual/360. and 30/360.
        settlement (datetime, optional): The settlement date. Defaults to today.
        freq (int, optional):
        Coupon frequency per year: 1 (annual). 2 (semi-annual)
                                    4 (quarterly), 12 (monthly)
                                    Defaults to 2.
    """
    from datetime import date
    import calendar
    from dateutil.relativedelta import relativedelta
    #Validate Data
    maturity = validate_date(maturity)

    if settlement is None:
        settlement = date.today()
    else:
        settlement = validate_date(settlement)

    if freq not in [1, 2, 4, 12]:
        print(f"⚠️ Warning: Freq {freq} invalid. Assumed Semi-Annual (2).")
        freq = 2

    try:
        coupon = float(coupon)
        if coupon < 0: raise ValueError
    except:
        raise ValueError("Coupon must be a positive number.")

    # Find the Correct Coupon Period
    # Get all the bond's payment dates
    pay_dates = scheduled_pay_dates(maturity, settlement=settlement, freq=freq)

    # Should sorted but check
    pay_dates.sort()

    # The first date after the settlement date is the next coupon date
    next_coupon = None
    for d in pay_dates:
        if d >= settlement:
            next_coupon = d
            break

    # Bond has matured or annual coupon is zero
    if next_coupon is None or coupon==0:
        return 0.0

    #Calculate Previous Coupon Date
    num_months = int(12 // freq)
    prev_coupon = next_coupon - relativedelta(months=num_months)

    # Check for Month End adjustment on the calculated previous date
    is_next_month_end = next_coupon.day == calendar.monthrange(maturity.year, maturity.month)[1]

    if is_next_month_end:
        last_day_of_prev_month = calendar.monthrange(prev_coupon.year, prev_coupon.month)[1]
        prev_coupon = date(prev_coupon.year, prev_coupon.month, last_day_of_prev_month)

    # In Python date math, (Settlement - Prev) automatically excludes the end date,
    # which correctly represents "days held".

    accrued_value = 0.0

    if day_type == 'Actual/Actual':
        days_since_last = (settlement - prev_coupon).days
        days_between = (next_coupon - prev_coupon).days
        # Formula: Coupon/Freq * (DaysHeld / DaysInPeriod)
        accrued_value = (coupon / freq) * (days_since_last / days_between)

    elif day_type == '30/360':
        days_since_last = _days_30_360(prev_coupon, settlement)
        # Formula: Coupon * (Days360 / 360)
        accrued_value = coupon * (days_since_last / 360.0)

    elif day_type == 'Actual/360':
        days_since_last = (settlement - prev_coupon).days
        accrued_value = coupon * (days_since_last / 360.0)

    elif day_type == 'Actual/365':
        days_since_last = (settlement - prev_coupon).days
        accrued_value = coupon * (days_since_last / 365.0)

    else:
        # Fallback
        print(f"⚠️ Warning: Unknown day_type {day_type}. Using Actual/Actual.")
        days_since_last = (settlement - prev_coupon).days
        days_between = (next_coupon - prev_coupon).days
        accrued_value = (coupon / freq) * (days_since_last / days_between)

    return accrued_value

## Calculating accrued interest for <font color='green'>five_bonds</font> DataFrame

In [13]:
maturities=five_bonds.index
coupons=five_bonds['Coupon']
settlement=date(2025,1,21)
freq=2
for maturity,coupon in zip(maturities,coupons):
  bond_accrued_interest=accrued_interest(maturity,coupon,
                                         settlement=settlement,
                                         freq=2)
  display(f'Maturity {maturity}---Coupon {coupon}--\
  Accrued Interest {round(bond_accrued_interest,3)}')

'Maturity 2025-02-28 00:00:00---Coupon 1.125--  Accrued Interest 0.444'

'Maturity 2025-07-15 00:00:00---Coupon 3.0--  Accrued Interest 0.05'

'Maturity 2025-08-31 00:00:00---Coupon 0.25--  Accrued Interest 0.099'

'Maturity 2026-01-15 00:00:00---Coupon 3.875--  Accrued Interest 0.064'

'Maturity 2028-02-29 00:00:00---Coupon 4.0--  Accrued Interest 1.587'

## Using the <font color='green'>apply</font> method to add accrured interst to the <font color='green'>bond_data</font> (304 bonds) DataFrame.

The <font color='green'>apply</font> method eliminates the need to iterate through all the values of a row or column. While many examples can be vectorized by operating directly on columns, <font color='green'>apply</font> is particularly useful for complex scenarios, such as calculating accrued interest, where column-wise operations are not feasible.$^{5}$

The benefits of using <font color='green'>apply</font>  include:

1. **Direct Column Access:** The lambda function can directly access specific columns within each row.  
2. **Simplified Code:** It leads to less complex and more easily understandable code.

The method is demonstrated with the DataFrame of Chapter One, <font color='green'>rows_to_columns</font>.  The examples could be solved with column operations, but their simplicity makes the use of <font color='green'>apply</font> clear.

```
rows_to_columns=pd.DataFrame({'Rates':[0.03,0.05,0.07],
                              'Future Values':[1.030455,1.05171,1.072508],
                              'Present Value':[0.970446,0.951229,0.93294]})
display(rows_to_columns)
rows_to_columns['Discrete']= rows_to_columns.apply(
    lambda row: np.log(1+row['Rates']),
    axis=1)
display(rows_to_columns)
rows_to_columns['Log Future Values']=rows_to_columns['Future Values'].apply(
    lambda value: np.log(value))
display(rows_to_columns)
```
$^{5}$ [Pandas method apply](https://patrickjhess.github.io/Introduction-To-Python-For-Financial-Python/An_Introduction_To_Pandas.html#the-apply-method)


In [ ]:
# Create rows_to_columns DataFrame
rows_to_columns=pd.DataFrame({'Rates':[0.03,0.05,0.07],
                              'Future Values':[1.030455,1.05171,1.072508],
                              'Present Value':[0.970446,0.951229,0.93294]})
display(rows_to_columns)
# Access the value of rows
rows_to_columns['Discrete']= rows_to_columns.apply(
    lambda row: np.log(1+row['Rates']),
    axis=1)
display(rows_to_columns)
# Access the values of column
rows_to_columns['Log Future Values']=rows_to_columns['Future Values'].apply(
    lambda value: np.log(value))
display(rows_to_columns)

Rates  Future Values  Present Value
0   0.03       1.030455       0.970446
1   0.05       1.051710       0.951229
2   0.07       1.072508       0.932940

Rates  Future Values  Present Value  Discrete
0   0.03       1.030455       0.970446  0.029559
1   0.05       1.051710       0.951229  0.048790
2   0.07       1.072508       0.932940  0.067659

Rates  Future Values  Present Value  Discrete  Log Future Values
0   0.03       1.030455       0.970446  0.029559           0.030000
1   0.05       1.051710       0.951229  0.048790           0.050417
2   0.07       1.072508       0.932940  0.067659           0.070000

## Update <font color='green'>bond_data</font> with accrued interest
 The DataFrame <font color='green'>bond_data</font> is expanded by applying the <font color='green'>accurred_interesta</font> function to each row of <font color='green'>bond_data</font>.


```
bond_data['Accrued'] = bond_data.apply(
    lambda row: accrued_interest(
        maturity=row.name,      # The row's index label (maturity date) is accessed via .name
        coupon=row['Coupon'],   # Access the 'Coupon' value from the row
        settlement=settlement   # Use the globally defined settlement date
    ), axis=1)
```



In [ ]:
settlement=date(2025,1,21)
# Use DataFrame.apply() to create the new column
# 'axis=1' tells pandas to apply the function to each row.
# A lambda function is used to pass the correct columns from each row to accrued_interest.
bond_data['Accrued'] = bond_data.apply(
    lambda row: accrued_interest(
        maturity=row.name,      # The row's index label (maturity date) is accessed via .name
        coupon=row['Coupon'],   # Access the 'Coupon' value from the row
        settlement=settlement   # Use the globally defined settlement date
    ), axis=1)

bond_data

Description  Coupon  \
Maturity Date                                                              
2025-01-28     UNITED STATES TREAS BILLS ZERO CPN       0.000...   0.000   
2025-01-30     UNITED STATES TREAS BILLS ZERO CPN       0.000...   0.000   
2025-01-31     UNITED STATES TREAS SER U-2025          1.3750...   1.375   
2025-01-31     UNITED STATES TREAS SER AW-2025         4.1250...   4.125   
2025-01-31     UNITED STATES TREAS SER G-2025          2.5000...   2.500   
...                                                          ...     ...   
2029-11-30     UNITED STATES TREAS SER AG-2029         4.1250...   4.125   
2029-11-30     UNITED STATES TREAS SER S-2029          3.8750...   3.875   
2029-12-31     UNITED STATES TREAS SER T-2029          3.8750...   3.875   
2029-12-31     UNITED STATES TREAS SER AH-2029         4.3750...   4.375   
2030-01-31     UNITED STATES TREAS SER G-2030          3.5000...   3.500   

               Price Bid  Price Ask  Bid Size  Ask Size   Accrued  
Maturity Date                                                      
2025-01-28        99.929     99.930    100000    100000  0.000000  
2025-01-30        99.906     99.907     40000     40000  0.000000  
2025-01-31        99.921     99.934     60000     60000  0.650136  
2025-01-31        99.988     99.997     60000     60000  1.950408  
2025-01-31        99.953     99.965     60000     60000  1.182065  
...                  ...        ...       ...       ...       ...  
2029-11-30        98.910     98.914    100000    100000  0.589286  
2029-11-30        97.763     97.782     40000     40000  0.553571  
2029-12-31        97.734     97.738     65000     65000  0.234203  
2029-12-31        99.988     99.989      7000      7000  0.264423  
2030-01-31        95.994     96.011     16000     16000  1.654891  

[304 rows x 7 columns]

## <font color='green'>Application:Calculate A Transaction Price</font>

<div style="
    border-left: 12px solid green;
    line-height: 1.5;
    padding: 15px">
<br>

Assume a bond matures on January 15$^{th}$ 2030 and the settlement date is May 1$^{st}$ 2024. The bond's clean quoted price is $95 and the coupon is 5.0. see [Chapter Two Hints:Calculate A Transaction Price](https://patrickjhess.github.io/Hints-Results/Chapter_Two_Hints.html#calculate-a-transaction-price), and check the [expected results here](https://patrickjhess.github.io/Hints-Results/Chapter_Two_Results.html#calculate-a-transaction-price).

<br>
</div>


## Creating The Workbook
 The <font color='green'>create_workbook</font> function transforms the bond_data DataFrame into an Excel workbook. It takes the sheet name, DataFrame name, and the dictionary <font color='green'>save_config</font>  as arguments. The sheet and DataFrame names are required, while save_config defaults to an empty dictionary. When <font color='green'>save_config</font> is not specified, the workbook is created in the current working directory of the notebook.$^{6}$

Note: The working directory for a Jupyter notebook is its current folder. For a Colab notebook, the location is 'contents' and is only available during runtime.

$^{6}$ The <font color='green'>create_workbook</font> function utilizes the <font color='green'>save_results</font> function located in the <font color='green'>basic_concepts_fixed_income</font> module.

In [ ]:
save_config=dict(volume='Basic Concepts Of Fixed Income',
                 chapter='Accrued Interest',
                 file_name='Accrued Interest.xlsx')
create_workbook('Basic Bond Data',bond_data,save_config)

❓ Do you want to save the file? (y/n) (press enter for n): n


###***❌ File Not Saved***.

### ***✅ Successfully wrote and formatted sheet Basic Bond Data in Accrued Interest.xlsx***

## <font color='green'>Application:Create An Excel Workbook</font>

<div style="
    border-left: 12px solid green;
    line-height: 1.5;
    padding: 15px">
<br>

Use the first ten rows of bond_data to create the workbook 'Test Data' and save it to the worksheet 'Ten Rows Of Data' see [Chapter Two Hints: Create A Workbook](https://patrickjhess.github.io/Hints-Results/Chapter_Two_Hints.html#create-a-workbook), and check the [expected results here](https://patrickjhess.github.io/Hints-Results/Chapter_Two_Results.html#create-a-workbook).

<br>
</div>


## <span style="text-align: left; color:green; font-family: 'Franklin Gothic Medium', sans-serif; margin-top: 1.0em; margin-bottom: 0em; font-style: italic;">Chapter Exercise</span>
<span style="text-align: left; color:green; font-family: 'Franklin Gothic Medium', sans-serif; margin-top: 0; margin-bottom: 0.5em; font-style: italic;"><big><font color='black'>Create A Workbook With Accrued Interest For 100 Bonds</big></font></span>

<span style="display:block;
    border-left: 12px solid green;
    font-family: 'Garamond', serif;
    line-height: 1.5;
    padding: 15px">
<big>Using Download bond_jan_21_2025.xlsx For The Exercoise</big><br></span>
<span style="display:block;
    border-left: 12px solid green;
    font-family: 'Garamond', serif;
    line-height: 2;">
<br><big>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.&nbsp;For the first 100 rows of the workbook calculate the accrued interest and save it in a new DataFrame.<br><br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2.&nbsp;Save the new DataFrame as an Excel Workbook..</big><br></span>
<span style="display:block;
    border-left: 12px solid green;
    font-family: 'Garamond', serif;
    line-height: 1.5;
    padding: 15px">
<br><big>For hints, see [Chapter One Hints](https://patrickjhess.github.io/Hints-Results/Chapter_One_Hints.html#chapter-one-exercise). For expected results, see [Chapter One Expected Results](https://patrickjhess.github.io/Hints-Results/Chapter_One_Results.html#chapter-one-exercise).</big>

</span>